subsys/net/ip/tcp2.c

```diff
@@ -1054,18 +1054,26 @@ static void tcp_timewait_timeout(struct k_work *work)
 {
 	struct tcp *conn = CONTAINER_OF(work, struct tcp, timewait_timer);
 
+	k_mutex_lock(&tcp_lock, K_FOREVER);
+
 	NET_DBG("conn: %p %s", conn, log_strdup(tcp_conn_state(conn, NULL)));
 
 	/* Extra unref from net_tcp_put() */
 	net_context_unref(conn->context);
+
+	k_mutex_unlock(&tcp_lock);
```

Comment:
I'm not sure what this lock is going to fix/prevent. There is already a per-context lock in `net_context_unref()` and then `k_mutex_lock(&tcp_lock)` (which will be locked recursively second time) in `tcp_conn_unref()`.